File: rubocop/tapify.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

$stdout.sync = true
$stderr.sync = true

require 'json'
require 'yaml'

def annotation_level(level)
  case level
  when 'error', 'fatal' then 'failure'
  when 'warning' then 'warning'
  else 'notice'
  end
end

report = JSON.parse($stdin.read)

files = report.fetch('files')

$stdout.puts('--- TAP')
$stdout.puts("1..#{files.size}")

files.each_with_index do |data, i|
  path = data.fetch('path')
  offenses = data.fetch('offenses')

  if offenses.empty?
    $stdout.puts("ok #{i + 1} - #{path}")
  else
    $stdout.puts("not ok #{i + 1} - #{path}")
    annotations = offenses.map do |offense|
      {
        'path' => path,
        'start_line' => offense.fetch('location').fetch('line'),
        'end_line' => offense.fetch('location').fetch('line'),
        'annotation_level' => annotation_level(offense.fetch('severity')),
        'message' => offense.fetch('message'),
        'title' => "Rubocop #{offense.fetch('cop_name')}"
      }
    end
    yaml = YAML.dump(annotations).lines.map do |line|
      "  #{line}" # prepend two spaces for tap correctness
    end
    yaml << '  ...'
    $stdout.puts(yaml.join)
  end
end

$stdout.puts('--- TAP')

File: cfnlint/tapify.py
#!/usr/bin/env python

import yaml, json, sys

def annotation_level(level):
    levels = {
        'Error': 'failure',
        'Warning': 'warning'
    }

    return levels.get(level, 'notice')

def main():
    report = json.load(sys.stdin)

    print('--- TAP')
    print('1..{}'.format(len(report)))

    for i, data in enumerate(report):
        path = data.get('Filename')
        message = data.get('Message')

        print('not ok {} - {}'.format(i + 1, path))
        encoded = yaml.dump({
            'path': path,
            'start_line': data.get('Location').get('Start').get('LineNumber'),
            'end_line': data.get('Location').get('End').get('LineNumber'),
            'annotation_level': annotation_level(data.get('Level')),
            'message': message,
            'title': 'cfn-lint: {}'.format(data.get('Rule').get('Id'))
        }, explicit_start = True, explicit_end = True, default_flow_style = False)

        for line in encoded.splitlines():
            print('  {}'.format(line))

    print('--- TAP')

if __name__ == "__main__":
    main()
